File: backend/app/schemas/community.py
"""Contracts for the opt-in anonymous peer support surface."""
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


ReactionType = Literal["support", "same_path", "learned"]


class CommunityPostCreate(BaseModel):
    source_type: Literal["diet_record"]
    source_id: int = Field(gt=0)
    caption: Optional[str] = Field(default=None, max_length=280)
    idempotency_key: str = Field(min_length=8, max_length=160)


class CommunityReactionUpdate(BaseModel):
    reaction: ReactionType


class CommunityReportCreate(BaseModel):
    reason: str = Field(min_length=2, max_length=200)


class CommunityPostResponse(BaseModel):
    id: int
    anonymous_name: str
    source_type: str
    snapshot: dict[str, Any]
    caption: Optional[str]
    status: str
    reaction_counts: dict[str, int]
    my_reaction: Optional[ReactionType]
    is_owner: bool
    created_at: datetime


class CommunityFeedResponse(BaseModel):
    items: list[CommunityPostResponse]


class CommunityReportResponse(BaseModel):
    report_count: int
    status: str

File: mobile/services/community.ts
import api from './api';

export type CommunityReaction = 'support' | 'same_path' | 'learned';

export interface CommunityDietSnapshot {
  meal_type: string;
  record_date: string;
  food_items: string;
  calories: number | null;
  protein: number | null;
  carbs: number | null;
  fat: number | null;
  fiber: number | null;
}

export interface CommunityPost {
  id: number;
  anonymous_name: string;
  source_type: 'diet_record';
  snapshot: CommunityDietSnapshot;
  caption: string | null;
  status: 'active' | 'under_review' | 'deleted';
  reaction_counts: Record<CommunityReaction, number>;
  my_reaction: CommunityReaction | null;
  is_owner: boolean;
  created_at: string;
}

export async function listCommunityPosts({
  limit = 20,
  beforeId,
}: {
  limit?: number;
  beforeId?: number;
} = {}): Promise<CommunityPost[]> {
  const { data } = await api.get<{ items: CommunityPost[] }>('/community/posts', {
    params: {
      limit,
      ...(beforeId ? { before_id: beforeId } : {}),
    },
  });
  return data.items;
}

export async function getCommunityPostForDietRecord(
  recordId: number,
): Promise<CommunityPost | null> {
  try {
    const { data } = await api.get<CommunityPost>(
      `/community/posts/source/diet_record/${recordId}`,
    );
    return data;
  } catch (error: any) {
    if (error?.response?.status === 404) return null;
    throw error;
  }
}

export async function publishDietRecordToCommunity(
  recordId: number,
  caption: string,
  idempotencyKey: string,
): Promise<CommunityPost> {
  const { data } = await api.post<CommunityPost>('/community/posts', {
    source_type: 'diet_record',
    source_id: recordId,
    caption: caption.trim() || undefined,
    idempotency_key: idempotencyKey,
  });
  return data;
}

export async function setCommunityReaction(
  postId: number,
  reaction: CommunityReaction,
): Promise<CommunityPost> {
  const { data } = await api.put<CommunityPost>(`/community/posts/${postId}/reaction`, { reaction });
  return data;
}

export async function removeCommunityReaction(postId: number): Promise<CommunityPost> {
  const { data } = await api.delete<CommunityPost>(`/community/posts/${postId}/reaction`);
  return data;
}

export async function deleteCommunityPost(postId: number): Promise<void> {
  await api.delete(`/community/posts/${postId}`);
}

export async function reportCommunityPost(
  postId: number,
  reason: string,
): Promise<{ report_count: number; status: string }> {
  const { data } = await api.post<{ report_count: number; status: string }>(
    `/community/posts/${postId}/report`,
    { reason },
  );
  return data;
}
